# DS Automation Assignment

Using our prepared churn data from week 2:
- use pycaret to find an ML algorithm that performs best on the data
    - Choose a metric you think is best to use for finding the best model; by default, it is accuracy but it could be AUC, precision, recall, etc. The week 3 FTE has some information on these different metrics.
- save the model to disk
- create a Python script/file/module with a function that takes a pandas dataframe as an input and returns the probability of churn for each row in the dataframe
    - your Python file/function should print out the predictions for new data (new_churn_data.csv)
    - the true values for the new data are [1, 0, 0, 1, 0] if you're interested
- test your Python module and function with the new data, new_churn_data.csv
- write a short summary of the process and results at the end of this notebook
- upload this Jupyter Notebook and Python file to a Github repository, and turn in a link to the repository in the week 5 assignment dropbox

*Optional* challenges:
- return the probability of churn for each new prediction, and the percentile where that prediction is in the distribution of probability predictions from the training dataset (e.g. a high probability of churn like 0.78 might be at the 90th percentile)
- use other autoML packages, such as TPOT, H2O, MLBox, etc, and compare performance and features with pycaret
- create a class in your Python module to hold the functions that you created
- accept user input to specify a file using a tool such as Python's `input()` function, the `click` package for command-line arguments, or a GUI
- Use the unmodified churn data (new_unmodified_churn_data.csv) in your Python script. This will require adding the same preprocessing steps from week 2 since this data is like the original unmodified dataset from week 1.

# Assignment

## Import Library and Data

In [1]:
import pandas as pd

from pycaret.classification import setup, compare_models, pull, save_model, load_model, evaluate_model, predict_model

In [2]:
# load churn data
df = pd.read_csv('data/churn_data_modified.csv')

In [3]:
df.head()

,customerID,tenure,PhoneService,Contract,PaymentMethod,MonthlyCharges,TotalCharges,charge_per_tenure,Churn
0,7590-VHVEG,1,0,0,0,29.85,29.85,29.850000,0
1,5575-GNVDE,34,1,1,1,56.95,1889.50,55.573529,0
2,3668-QPYBK,2,1,0,1,53.85,108.15,54.075000,1
3,7795-CFOCW,45,0,1,2,42.30,1840.75,40.905556,0
4,9237-HQITU,2,1,0,0,70.70,151.65,75.825000,1


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7032 entries, 0 to 7031
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   customerID         7032 non-null   object 
 1   tenure             7032 non-null   int64  
 2   PhoneService       7032 non-null   int64  
 3   Contract           7032 non-null   int64  
 4   PaymentMethod      7032 non-null   int64  
 5   MonthlyCharges     7032 non-null   float64
 6   TotalCharges       7032 non-null   float64
 7   charge_per_tenure  7032 non-null   float64
 8   Churn              7032 non-null   int64  
dtypes: float64(3), int64(5), object(1)
memory usage: 494.6+ KB


## AutoML with pycaret

In [5]:
automl = setup(df, target='Churn', session_id=123)

,Description,Value
0,Session id,123
1,Target,Churn
2,Target type,Binary
3,Original data shape,"(7032, 9)"
4,Transformed data shape,"(7032, 9)"
5,Transformed train set shape,"(4922, 9)"
6,Transformed test set shape,"(2110, 9)"
7,Numeric features,7
8,Categorical features,1
9,Preprocess,True


## Compare Models

In [6]:
best_model = compare_models()

,Model,Accuracy,AUC,Recall,Prec.,F1,Kappa,MCC,TT (Sec)
knn,K Neighbors Classifier,0.7666,0.7462,0.4358,0.5815,0.4978,0.3499,0.3563,0.0150
nb,Naive Bayes,0.7582,0.8153,0.6400,0.5389,0.5847,0.4161,0.4194,0.0090
svm,SVM - Linear Kernel,0.7497,0.7435,0.5092,0.5868,0.5146,0.3583,0.3797,0.0100
lr,Logistic Regression,0.7345,0.8318,0.0008,0.1000,0.0015,0.0011,0.0075,0.3410
dt,Decision Tree Classifier,0.7343,0.5000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0080
ridge,Ridge Classifier,0.7343,0.8219,0.0000,0.0000,0.0000,0.0000,0.0000,0.0080
rf,Random Forest Classifier,0.7343,0.6951,0.0000,0.0000,0.0000,0.0000,0.0000,0.0400
qda,Quadratic Discriminant Analysis,0.7343,0.5000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0070
ada,Ada Boost Classifier,0.7343,0.5000,0.0000,0.0000,0.0000,0.0000,0.0000,0.0080
gbc,Gradient Boosting Classifier,0.7343,0.5199,0.0000,0.0000,0.0000,0.0000,0.0000,0.0340


In [7]:
best_model

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=-1, n_neighbors=5, p=2,
                     weights='uniform')

In [8]:
evaluate_model(best_model)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [9]:
df.iloc[-2:-1].shape

(1, 9)

## Save the Model

In [10]:
save_model(best_model, 'best_churn_model')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['tenure', 'PhoneService',
                                              'Contract', 'PaymentMethod',
                                              'MonthlyCharges', 'TotalCharges',
                                              'charge_per_tenure'],
                                     transformer=SimpleImputer(add_indicator=False,
                                                               copy=True,
                                                               fill_value=None,
                                                               keep_empty_features=False,
                                                               missing_values=nan,
                                                               strategy='mean'))),
                 ('c...
                                     transformer=TargetEn

In [11]:
import pickle

with open('best_churn_model.pk', 'wb') as f:
    pickle.dump(best_model, f)

In [12]:
with open('best_churn_model.pk', 'rb') as f:
    loaded_model = pickle.load(f)

## Test Prediction Script

In [13]:
from IPython.display import Code

Code('predict_churn.py')

import pandas as pd
from pycaret.classification import load_model, predict_model

def predict_churn_probability(dataframe: pd.DataFrame) -> pd.DataFrame:
    """
    Loads the trained churn prediction model and predicts churn probabilities
    for the input DataFrame.

    Args:
        dataframe: Pandas DataFrame with the same structure as the training data
                   (excluding the target variable).

    Returns:
        Pandas DataFrame with 'customerID' and 'Probability of Churn' columns.
    """
    # Load the trained model
    loaded_model = load_model('best_churn_model', verbose=False)

    # # Predict churn probabilities
    predictions = loaded_model.predict_proba(dataframe)

    # Convert the predictions array to a DataFrame
    predictions_df = pd.DataFrame(predictions, columns=['Probability_0', 'Probability_1'])

    # Return the churn probability (probability of class 1)
    return predictions_df[['Probability_1']]


if __name__ == '__main__':
    try:
        # Load the new data (which already contains 'charge_per_tenure' based on inspection)
        new_data = pd.read_csv('data/new_churn_data.csv')
        churn_probabilities = predict_churn_probability(new_data)
        print("Churn Probabilities for new data:")
        print(churn_probabilities)
    except FileNotFoundError:
        print("Error: new_churn_data.csv not found. Please make sure the file is in the correct path.")

In [14]:
%run predict_churn.py

Churn Probabilities for new data:
   Probability_1
0            0.0
1            0.4
2            0.0
3            0.0
4            0.6


<Figure size 800x550 with 0 Axes>

In this assignment, we used the prepared churn data from Week 2 to build a machine learning model for churn prediction using the PyCaret library.

We started by setting up the PyCaret environment and comparing various classification models to identify the best performer based on appropriate metrics. After training and evaluating the models, the KNeighborsClassifier was selected as the best model.

The best model was then saved to disk for later use. A Python script named predict_churn.py was created with a function predict_churn_probability that loads the saved model and predicts the probability of churn for new data provided as a pandas DataFrame.

Finally, the predict_churn.py script was tested with the new_churn_data.csv file. The script successfully loaded the data, applied the necessary preprocessing (handled by the loaded PyCaret pipeline by passing only the original features), and output the predicted churn probabilities for each customer in the new dataset.

Overall, the process successfully demonstrated how to use an automated machine learning library like PyCaret to build, save, and deploy a churn prediction model for making predictions on new data.

## Optional Tasks

### Calculate prediction percentiles

In [15]:
loaded_model = load_model('best_churn_model', verbose=False)

In [16]:
# Get predictions on the training data
# train_predictions = predict_model(loaded_model, data=df)
df_to_predict = df[df.columns[:-1]]

train_predictions = loaded_model.predict_proba(df_to_predict)

In [17]:
from scipy.stats import percentileofscore

train_predictions = loaded_model.predict_proba(df_to_predict)

# Extract churn probabilities using the correct column name
train_churn_probabilities = pd.DataFrame(train_predictions, columns=['Probability_0', 'Probability_1'])

# Calculate percentiles for the training data churn probabilities
train_percentiles = train_churn_probabilities.apply(lambda x: percentileofscore(train_churn_probabilities, x))
train_percentiles.columns = ['Percentile_0', 'Percentile_1']

# Store the training data churn probabilities and their percentiles
train_probability_distribution = pd.concat([train_churn_probabilities['Probability_1'], train_percentiles['Percentile_1']], axis=1)
train_probability_distribution.columns = ['Churn Probability', 'Probability Percentile']

display(train_probability_distribution.head())
display(train_probability_distribution.describe())

,Churn Probability,Probability Percentile
0,0.4,0.014221
1,0.0,0.014221
2,0.6,0.028441
3,0.0,0.014221
4,0.6,0.028441


,Churn Probability,Probability Percentile
count,7032.000000,7032.000000
mean,0.265074,0.017198
std,0.283742,0.005786
min,0.000000,0.014221
25%,0.000000,0.014221
50%,0.200000,0.014221
75%,0.400000,0.014221
max,1.000000,0.028441


In [18]:
train_probability_distribution_path = 'train_probability_distribution.csv'
train_probability_distribution.to_csv(train_probability_distribution_path, index=False)
print(f"Training probability distribution saved to {train_probability_distribution_path}")

Training probability distribution saved to train_probability_distribution.csv


### Test prediction percentiles

In [19]:
from IPython.display import Code

Code('predict_churn_prob.py')

import pandas as pd
from pycaret.classification import load_model, predict_model
from scipy.stats import percentileofscore
import sys

class ChurnPredictor:
    """
    A class to predict customer churn probability and percentile.
    """
    def __init__(self, model_path='best_churn_model', train_prob_dist_path='train_probability_distribution.csv'):
        """
        Initializes the ChurnPredictor by loading the model and training data distribution.

        Args:
            model_path: Path to the saved PyCaret model.
            train_prob_dist_path: Path to the CSV file containing the training
                                 probability distribution ('Churn Probability' and 'Probability Percentile').
        """
        self.loaded_model = load_model(model_path, verbose=False)
        try:
            self.train_probability_distribution = pd.read_csv(train_prob_dist_path)
        except FileNotFoundError:
            print(f"Error: Training probability distribution file not found at {train_prob_dist_path}")
            print("Please ensure 'train_probability_distribution.csv' is in the same directory or provide the correct path.")
            sys.exit(1) # Exit if the training distribution file is not found


    def preprocess_data(self, dataframe: pd.DataFrame) -> pd.DataFrame:
        """
        Applies the necessary preprocessing steps to the input DataFrame
        to match the training data structure.

        Args:
            dataframe: Input DataFrame (in the format of new_unmodified_churn_data.csv).

        Returns:
            Preprocessed DataFrame ready for prediction.
        """
        # Make a copy to avoid modifying the original DataFrame
        df_processed = dataframe.copy()

        # Drop rows with missing TotalCharges (matching the training data preprocessing)
        df_processed.dropna(subset=['TotalCharges'], inplace=True)

        # Convert 'PhoneService' column
        df_processed['PhoneService'] = df_processed['PhoneService'].apply(lambda x: 1 if x == 'Yes' else 0)

        # Convert 'Contract' column using mapping
        contract_mapping = {'Month-to-month': 0, 'One year': 1, 'Two year': 2}
        df_processed['Contract'] = df_processed['Contract'].map(contract_mapping)

        # Convert 'PaymentMethod' column using mapping
        payment_mapping = {'Electronic check': 0, 'Mailed check': 1, 'Bank transfer (automatic)': 2, 'Credit card (automatic)': 3}
        df_processed['PaymentMethod'] = df_processed['PaymentMethod'].map(payment_mapping)

        # Calculate 'charge_per_tenure' column
        # Handle potential division by zero if tenure is 0
        df_processed['charge_per_tenure'] = df_processed.apply(
            lambda row: row['TotalCharges'] / row['tenure'] if row['tenure'] > 0 else 0, axis=1
        )


        # Reorder columns to match the training data used by the model,
        # excluding the target 'Churn' column if present in the input.
        expected_columns = ['customerID', 'tenure', 'PhoneService', 'Contract', 'PaymentMethod',
                            'MonthlyCharges', 'TotalCharges', 'charge_per_tenure']

        # Drop columns not in the expected list (like 'Churn' if it exists in new data)
        extra_cols = [col for col in df_processed.columns if col not in expected_columns]
        df_processed.drop(columns=extra_cols, inplace=True)

        # Ensure all expected columns are present, if not, add them with a default value (e.g., 0)
        for col in expected_columns:
            if col not in df_processed.columns:
                df_processed[col] = 0 # Or handle this case based on domain knowledge

        # Reorder the columns
        df_processed = df_processed[expected_columns]

        return df_processed

    def predict_churn_probability_and_percentile(self, dataframe: pd.DataFrame) -> pd.DataFrame:
        """
        Predicts churn probabilities and calculates their percentiles within
        the training data distribution for the input Dat

In [25]:
%run predict_churn_prob.py


Churn Predictions with Percentiles:
   Probability of Churn  Probability Percentile
0                   0.0               19.340159
1                   0.4               71.053754
2                   0.0               19.340159
3                   0.0               19.340159
4                   0.6               84.620307


We built a single Python script called predict_churn_prob.py that defines a ChurnPredictor class to handle everything: 
* It asks (via input()) for the path to your raw new_unmodified_churn_data.csv, then it loads that file
* Applies exactly the preprocessing steps—missing-value imputation, categorical encoding, scaling, feature interactions—loads.
* The churn probability provides where score falls in the training-set probability distribution.

### h2o

In [21]:
import h2o
from h2o.frame import H2OFrame

# Initialize H2O cluster
h2o.init()

Checking whether there is an H2O instance running at http://localhost:54321..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "24.0.2" 2025-07-15; OpenJDK Runtime Environment Zulu24.32+13-CA (build 24.0.2+12); OpenJDK 64-Bit Server VM Zulu24.32+13-CA (build 24.0.2+12, mixed mode, sharing)
  Starting server from /opt/anaconda3/envs/mynewenv/lib/python3.10/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /var/folders/t5/p1xqr73x1slfx2p_g4rrxkrh0000gn/T/tmp4y7ahsni
  JVM stdout: /var/folders/t5/p1xqr73x1slfx2p_g4rrxkrh0000gn/T/tmp4y7ahsni/h2o_ada_started_from_python.out
  JVM stderr: /var/folders/t5/p1xqr73x1slfx2p_g4rrxkrh0000gn/T/tmp4y7ahsni/h2o_ada_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.
Please download and install the latest version from: https://h2o-release.s3.amazonaws.com/h2o/latest_stable.html


H2O_cluster_uptime:,01 secs
H2O_cluster_timezone:,America/Toronto
H2O_data_parsing_timezone:,UTC
H2O_cluster_version:,3.46.0.7
H2O_cluster_version_age:,6 months
H2O_cluster_name:,H2O_from_python_ada_3hwryv
H2O_cluster_total_nodes:,1
H2O_cluster_free_memory:,5.984 Gb
H2O_cluster_total_cores:,10
H2O_cluster_allowed_cores:,10
H2O_cluster_status:,"locked, healthy"


In [22]:
# Load the churn data
h2o_df = h2o.import_file('data/churn_data.csv')

# Convert the target variable to a factor
h2o_df['Churn'] = h2o_df['Churn'].asfactor()

# Identify feature and target columns
target = 'Churn'
features = h2o_df.columns
features.remove(target)

print("H2OFrame loaded and target converted to factor.")
print("Features:", features)
print("Target:", target)

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%
H2OFrame loaded and target converted to factor.
Features: ['customerID', 'tenure', 'PhoneService', 'Contract', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges']
Target: Churn


In [23]:
from h2o.automl import H2OAutoML

# Initialize H2O AutoML
aml = H2OAutoML(max_runtime_secs=300, seed=42) # Set max_runtime_secs to 300 seconds (5 minutes)

# Train the AutoML model
aml.train(x=features, y=target, training_frame=h2o_df)

print("H2O AutoML training complete.")

AutoML progress: |
10:18:53.510: AutoML: XGBoost is not available; skipping it.
10:18:53.522: _train param, Dropping bad and constant columns: [customerID]


10:18:54.731: _train param, Dropping bad and constant columns: [customerID]

██
10:18:56.919: _train param, Dropping unused columns: [customerID]
10:18:57.177: _train param, Dropping bad and constant columns: [customerID]

██
10:18:59.332: _train param, Dropping bad and constant columns: [customerID]

█
10:18:59.977: _train param, Dropping bad and constant columns: [customerID]
10:19:00.638: _train param, Dropping bad and constant columns: [customerID]

█
10:19:01.304: _train param, Dropping unused columns: [customerID]
10:19:01.564: _train param, Dropping unused columns: [customerID]
10:19:01.848: _train param, Dropping bad and constant columns: [customerID]

██
10:19:03.289: _train param, Dropping bad and constant columns: [customerID]

██
10:19:04.43: _train param, Dropping bad and constant columns: [customerID]

██████
10:19:0

In [24]:
# Get the AutoML leaderboard
leaderboard = aml.leaderboard

print("H2O AutoML Leaderboard:")
leaderboard.head()

H2O AutoML Leaderboard:


model_id,auc,logloss,aucpr,mean_per_class_error,rmse,mse
StackedEnsemble_BestOfFamily_6_AutoML_1_20250928_101853,0.842858,0.418389,0.649354,0.241171,0.369501,0.136531
StackedEnsemble_BestOfFamily_4_AutoML_1_20250928_101853,0.842733,0.418422,0.649879,0.240866,0.369509,0.136537
StackedEnsemble_AllModels_4_AutoML_1_20250928_101853,0.841928,0.419021,0.650142,0.241084,0.369844,0.136785
StackedEnsemble_AllModels_3_AutoML_1_20250928_101853,0.84182,0.41926,0.64787,0.242012,0.369938,0.136854
GBM_grid_1_AutoML_1_20250928_101853_model_14,0.8418,0.420203,0.650355,0.240396,0.370177,0.137031
StackedEnsemble_AllModels_6_AutoML_1_20250928_101853,0.84122,0.419945,0.646379,0.237658,0.370335,0.137148
StackedEnsemble_Best1000_1_AutoML_1_20250928_101853,0.8411,0.420171,0.64668,0.244469,0.370305,0.137126
StackedEnsemble_BestOfFamily_3_AutoML_1_20250928_101853,0.840898,0.420393,0.647115,0.244051,0.370574,0.137325
StackedEnsemble_BestOfFamily_1_AutoML_1_20250928_101853,0.84073,0.420716,0.644472,0.242729,0.370736,0.137446
StackedEnsemble_AllModels_1_AutoML_1_20250928_101853,0.840573,0.420661,0.64856,0.23816,0.37066,0.137389


H2O’s stacked ensemble doutperforms PyCaret’s KNN on the churn dataset: the H2O model achieves an AUC of 0.8428 compared to KNN’s 0.7666. KNN excels in speed—training in just 0.014 seconds—making it ideal for quick prototyping, but it falls short on discrimination and probability calibration. H2O’s minutes-long training yields more reliable probability estimates and better handling of class imbalance, enabling threshold tuning, cost-sensitive learning, and robust deployment for retention campaigns.

# Summary

Write a short summary of the process and results here.

In this notebook, we performed the following steps:

1. **Loaded Data**: We loaded the churn data from a CSV file into a pandas DataFrame.
2. **PyCaret Setup and Model Training**: We used PyCaret to set up the experiment and automatically train and compare various classification models to predict customer churn. The best performing model (KNeighborsClassifier based on default metrics) was identified.
3. **Model Saving**: The best PyCaret model was saved to disk using `save_model`. We also demonstrated saving and loading the model using `pickle`.
4. **Churn Prediction Script**: A Python script (`predict_churn.py`) was created with a function to load the saved PyCaret model and predict churn probabilities for new data. This script was tested with `new_churn_data.csv`.
5. **Optional Task - Percentile Calculation**: We created churn prediction script (`predict_churn_prob.py`) that defines a ChurnPredictor class to calculate the percentile of each prediction within the distribution of churn probabilities from the training data. This script also included preprocessing steps for unmodified data and was tested with input file `new_churn_data_unmodified.csv`.
6. **H2O AutoML**: We explored another autoML library, H2O. We initialized an H2O cluster, loaded the data into an H2OFrame, converted the target variable, and trained an H2O AutoML model. The leaderboard of H2O models was displayed to show the performance of different algorithms. (Note: tried couple of other packages like TPOT, MLBOX but ran into confilct with version and\or confict with PyCaret)

## Results and Findings

Based on the analysis:

*   **PyCaret Model Performance**: The best model identified by PyCaret was the KNeighborsClassifier.
*   **Predictions**: The prediction scripts demonstrate how to load the saved PyCaret model and make predictions on new data, including calculating the percentile of the churn probability relative to the training data distribution.
*   **H2O AutoML Performance**: The H2O AutoML leaderboard shows that models generally performed the better based on the default sorting metric (AUC). The top performing H2O model achieved highesy AUC.
*   **Comparison**: Both PyCaret and H2O provided automated model training and comparison. PyCaret is good for quickly trying out various models with minimal code and provides interactive evaluation plots. H2O's AutoML also automates the process and includes advanced techniques like ensemble methods which often yield higher performance.

In summary, both PyCaret and H2O are powerful AutoML tools that can quickly identify well-performing models for a given task. The best model and its performance depend on the specific dataset and the chosen evaluation metric.